src/Npgsql/PoolManager.cs

```diff
@@ -13,20 +15,49 @@ namespace Npgsql;
 /// </remarks>
 static class PoolManager
 {
-    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools { get; } = new();
+    static ConcurrentDictionary<string, NpgsqlDataSource> _pools = new();
+
+    // Keep the IEnumerable API for testing purposes but remove the rest to avoid careless (unlocked) access to the
+    // private dictionary.
+    internal static IEnumerable<KeyValuePair<string, NpgsqlDataSource>> Pools => _pools.ToImmutableDictionary();
 
     internal static void Clear(string connString)
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        if (Pools.TryGetValue(connString, out var pool))
-            pool.Clear();
+        // ReSharper disable once InconsistentlySynchronizedField
+        var pools = _pools;
+        if (!pools.TryRemove(connString, out var pool))
+                return;
+
+        pool.Dispose();
+
+        // The same NpgsqlDataSource instance may be referenced via different connection strings
+        // in our dictionary, so we have to make sure that we remove all occurrences
+#if NET5_0_OR_GREATER
+        foreach (var pair in pools)
+            if (ReferenceEquals(pair.Value, pool))
+                pools.TryRemove(pair);
+#else
+        // Make sure nobody adds back a new pool with the same connection string
+        // as one we've already removed.
+        lock (pools)
```

Comment:
I think you need to take the lock above in the NET5_0_OR_GREATER code as well, no? Or just take it for the entire method, as proposed above.

Reply:
> I think you need to take the lock above in the NET5_0_OR_GREATER code as well, no? Or just take it for the entire method, as proposed above.

Not really. The `TryRemove()` overload that takes a `KeyValuePair<TKey,TValue>` in NET5_0_OR_GREATER checks both the key and the value for equality and doesn't remove anything if the value doesn't match. 
Since this already happens atomically in the `ConcurrentDictionary` we don't have to guard against accidental concurrent adding of a new value for the old key.

Reply:
True. But see my comment about generally locking for the entire methods.